File: packages/java/src/main/java/io/axene/mailer/Suppressions.java
package io.axene.mailer;

import com.fasterxml.jackson.databind.JavaType;

import java.util.Map;

/** Manage the do-not-send suppression list. Accessed as {@code client.suppressions()}. */
public final class Suppressions {

    private final ApiTransport transport;

    Suppressions(ApiTransport transport) {
        this.transport = transport;
    }

    /**
     * List suppressed addresses (paginated envelope; zero-based {@code page}).
     *
     * @param page   zero-based page index.
     * @param limit  page size (1-200).
     * @param search optional search filter, or null.
     * @return a page of suppressions.
     */
    public Page<Suppression> list(int page, int limit, String search) {
        String qs = Query.of().add("page", page).add("limit", limit).add("search", search).build();
        JavaType type = transport.mapper().getTypeFactory()
                .constructParametricType(Page.class, Suppression.class);
        return transport.request("GET", "/v1/suppressions" + qs, null, type);
    }

    /**
     * Suppress a single address. The {@code email} value is sent as the wire
     * field {@code email_address}.
     *
     * @param email  the address to suppress.
     * @param reason the suppression reason (defaults to {@code manual} when null).
     * @return the created suppression.
     */
    public Suppression add(String email, String reason) {
        Map<String, Object> body = Wire.map();
        body.put("email_address", email);
        body.put("reason", reason == null ? "manual" : reason);
        return transport.request("POST", "/v1/suppressions", body, transport.type(Suppression.class));
    }

    /**
     * Bulk-import suppressions from a file (one email per line).
     *
     * @param file     the raw file bytes.
     * @param filename the filename to advertise.
     * @return the bulk-import result.
     */
    public BulkSuppressionResult bulkUpload(byte[] file, String filename) {
        return transport.upload("/v1/suppressions/bulk", file, filename, transport.type(BulkSuppressionResult.class));
    }

    /**
     * Remove an address from the suppression list.
     *
     * @param id the suppression id.
     */
    public void remove(String id) {
        transport.request("DELETE", "/v1/suppressions/" + Query.enc(id), null, transport.type(Void.class));
    }
}

File: packages/ruby/lib/axene/mailer/resources/suppressions.rb
# frozen_string_literal: true

module Axene
  module Mailer
    module Resources
      # The +suppressions+ resource: manage the do-not-send list. Accessed as
      # +client.suppressions+.
      class Suppressions
        # @param transport [Axene::Mailer::Transport]
        def initialize(transport)
          @transport = transport
        end

        # List suppressed addresses (paginated envelope; zero-based +page+).
        #
        # @param page [Integer] default 0
        # @param limit [Integer] default 50
        # @param search [String, nil]
        # @return [Hash] { items:, total:, page:, limit: }
        def list(page: 0, limit: 50, search: nil)
          @transport.request(:get, "/v1/suppressions", query: { page: page, limit: limit, search: search })
        end

        # Suppress a single address. The +email+ argument maps to the wire field
        # +email_address+.
        #
        # @param email [String]
        # @param reason [String] default "manual"
        # @return [Hash]
        def add(email:, reason: "manual")
          @transport.request(:post, "/v1/suppressions", body: { email_address: email, reason: reason })
        end

        # Bulk-import suppressions from a file (one email per line). +file+ may be
        # raw bytes (a String) or a path to a readable file.
        #
        # @param file [String] raw bytes or a file path
        # @param filename [String]
        # @return [Hash] { added:, skipped:, total_processed: }
        def bulk_upload(file, filename: "suppressions.txt")
          bytes = read_file(file)
          @transport.upload("/v1/suppressions/bulk", bytes, filename)
        end

        # Remove an address from the suppression list.
        #
        # @param id [String]
        # @return [nil]
        def remove(id)
          @transport.request(:delete, "/v1/suppressions/#{Util.escape(id)}")
        end

        private

        def read_file(file)
          if file.is_a?(String) && file.length < 4096 && !file.include?("\n") && File.file?(file)
            File.binread(file)
          else
            file
          end
        end
      end
    end
  end
end
